Added the tensorflow complex64 data type.

AGaliciaMartinez opened the pull request:
Added a new support for complex 64 data type. The previous `TfTensor` class is now `_BaseTfTensor` from which the new two classes: `TfTensor128` and `TfTensor64` are derived. `_BasetTfTensor` wraps a tensorflow tensor of either `complex64` or `complex128` data type. Then the `TfTensor64` and `TfTensor128` are used as instances that the dispatchings system will know about. These only change the `_BaseTfTensor` class in that they force the init to be of the right data type. The specialisations will be adapted to work with both classes (add and expm have already been updated).

At this moment the following files are in its final form:
- convert.py
- tftensor.py
- add.py
- exmp.py

The rest of the specialisations will come once we decide whether this is the method to follow or not. 

The reason to have two classes is mainly due to how the dispatching system works. I have not been able to accommodate everything in one single class as I needed to provide two different classes to the dispatcher to be able to instantiate them from qobj individually. I wonder if this is the best approach for this problem though. 

TODO:
- [x] Update README.md with the new feature
- [x] Update example notebook.
- [x] Add the rest of the specialisations and their tests. (Add and Expm are working)
- [x] Add the test for convert and the new data type.
- [x] Add new dtypes to benchmarks.
- [x] Add warning when converting automatically from TfTensor128 to TfTensor64? I am not sure if it is a good idea to allow Add(TfTensor128, TfTensor64) to happen.
- [x] PEP8

AGaliciaMartinez committed:
Added the tensorflow64 data type.

AGaliciaMartinez committed:
Changed rtol for Expm64 tests.

hodgestar commented on src/qutip_tensorflow/core/data/add.py:
```diff
@@ -43,18 +43,20 @@ def iadd_tftensor(left, right, scale=1):
 qutip.data.add.add_specialisations(
     [
-        (TfTensor, TfTensor, TfTensor, add_tftensor),
+        (TfTensor128, TfTensor128, TfTensor128, add_tftensor),
```
Suggestion: If we write a function that adds the specializations, we can avoid duplicating everything and we can make it easier to register the specializations separately if for some reason someone wants one and not the other.

E.g.
```python
def _add_specialisations(cls):
    qutip.data.add.add_specialisations([
        (cls, cls, cls, add_tftensor),
    ])

_add_specialisations(TfTensor128)
_add_specialisations(TfTensor64)
```

hodgestar replied:
One downside is that the specialisations are currently added a bit all over the place (which is not necessarily a terrible thing), so the simple approach above will need a bit more thought.

AGaliciaMartinez replied:
I think I will stick with the simple approach here just because I only need to add one more specialisation each time. Although it may be necessary to consider such function if we want to bring support for multiple backends to qutip-tensornetwork.

hodgestar reviewed (commented):
Generally this approach looks good to me.

We should think a little bit more about how specialisations / data types are activated. It might be nice to be able to choose which to activate somehow in case one day there is a package with many options.

AGaliciaMartinez committed:
Renamed tests to use TfTensor128 by default.

This allows executing all tests but they still do not work.

AGaliciaMartinez committed:
Tests are now passing but only TfTensor128 is being tested.

AGaliciaMartinez committed:
Added convert tests for tftensor64

AGaliciaMartinez committed:
Added tests for both TfTenso128 and TfTensor64

This is done modifying the prefious tests to rely on the class
attributes Datatype and dtype that specifie the creation dtype and the
expected dtype of the wrapped tensor. In this way the tests for
TfTensor64 can be inherited from TfTensor128.

These tests required some minor modifications. The most notable one is
that some have to rely in comparisons between tf.tensors directly
instead of using the array method to compare numpy arrays. This si
because the array method always returns a complex128 dtype, regardless
of the underlying tf.tensor.

AGaliciaMartinez committed:
Added conj, adjoint and transpose specialisations.

AGaliciaMartinez committed:
Added specialisations for expect.

Tests do not require (surprisingly) to decrease tol. I was expecting it
to require a decrease due to the data type being complex 64 instead of
128.

AGaliciaMartinez commented:
> We should think a little bit more about how specialisations / data types are activated. It might be nice to be able to choose which to activate somehow in case one day there is a package with many options.

I think a way of defaulting to a data type would be very useful. For instance, I would like to make `TfTensor128` the default data type once qutip-tensorflow is imported, so that it entirely replaces the `Dense` data type.

AGaliciaMartinez committed:
Added Inner specialisation.

Quite suprisingly only inner needed adjustment for the rtol. Inner_op
did not.

AGaliciaMartinez committed:
Added the rest of the specialisations and tests for TfTensor64.

AGaliciaMartinez committed:
PEP 8

AGaliciaMartinez commented:
The tests are failing because im waiting for qutip/qutip#1723.

hodgestar commented on src/qutip_tensorflow/core/data/convert.py:
```diff
@@ -1,33 +1,62 @@
-def is_tftensor(data):
-    return isinstance(data, tf.Tensor)
+def _tf64_from_tf128(tftensor):
+    # Posibly raise a warning?
```
I guess this is only problematic if it happens accidentally (e.g. if the dispatcher chooses to use it for some reason).

hodgestar commented on src/qutip_tensorflow/core/data/convert.py:
```diff
@@ -1,33 +1,62 @@
 
 # User friendly name for conversion with `to` or Qobj creation functions:
-qutip.data.to.register_aliases(["tftensor", "TfTensor"], TfTensor)
+qutip.data.to.register_aliases(
+    ["tftensor128", "tftensor", "TfTensor"], TfTensor128
+)
+qutip.data.to.register_aliases(["tftensor64"], TfTensor64)
```
```suggestion
qutip.data.to.register_aliases(
    ["tftensor", "tftensor128", "TfTensor128"], TfTensor128
)
qutip.data.to.register_aliases(["tftensor64", "TfTensor64"], TfTensor64)
```
Just suggesting a slightly more consistent set of names -- so `tftensor` for the default, `tftensor128` and `tftensor64` for explicitly choosing, and then the two class names.

AGaliciaMartinez replied:
I agree, I will implement the changes.

hodgestar reviewed (approved):
I left some smaller comments, but I'm happy with the code.

We should also make the tests pass before merging. Can we do that without waiting for the QuTiP changes? It would be nice to decouple the two pieces of work if we can.

AGaliciaMartinez commented:
We can. I will do so :smile:.

AGaliciaMartinez committed:
Added suggestion from Simon.

Added TfTensor128 and TfTensor64 as strings to allow:
`qobj.to('TfTensor128')`

AGaliciaMartinez committed:
Changed test to not depend on PR in QuTiP.

I first modified the tests anticipating a PR I was going to propose in
QuTiP dev.major. Since this PR will not be merged, I reverted some of
the changes such that the work with the current dev.major.

AGaliciaMartinez committed:
Fixed test failing.

AGaliciaMartinez committed:
Updated README.

Included new `TfTensor64` and `TfTensor128` in the README.

AGaliciaMartinez committed:
Adapted benchmarks to unclude new data types.

AGaliciaMartinez committed:
Do not raise a warning.

I decided not to include a warning when changing from comple 128 to 64.

AGaliciaMartinez committed:
PEP 8

AGaliciaMartinez committed:
Deleted eigen file.

It mistakenly included it when committing black's changes.

AGaliciaMartinez commented:
@hodgestar I updated the readme and the benchmark script. Could you review those commits? After that  this should be ready to merge.

hodgestar reviewed (approved):
Looks good. Let's merge!

AGaliciaMartinez commented:
Perfect! Thank you for reviewing!
